### ISSUE TITLE
A sensible config.cache in our stdenv to speed up autoconf builds?

### ISSUE THREAD

**Issue by copumpkin** [NixOS/nixpkgs#14586]
## Issue description

All `autoconf` builds have a slow, non-parallelizable step at the beginning where every single build tries to rederive from scratch whether our word size is 4 bytes or 8, and so on. That step can take anywhere from a few seconds to over a minute, and is repeated once for most traditional unix tool builds.

[Sabotage Linux](http://sabotage.tech/) has a [config.cache](https://raw.githubusercontent.com/sabotage-linux/sabotage/master/KEEP/config.cache) they use in their build system to skip many of those checks, and I was wondering if we could use something similar in nixpkgs. Perhaps there's a way we could generate it once when we build a `stdenv`, and then have all builds reuse it?
### Steps to reproduce

Run any build.

**Comment by vcunat** [comment]
Years ago, we also had some experimental patches flying around that could parallelize the configure run somehow.

**Comment by copumpkin** [comment]
@vcunat that'd be nice, but doesn't feel like a widely supported mechanism. This seems like it's supported, and could save a massive amount of work.

At the same time, from [here](https://www.gnu.org/software/autoconf/manual/autoconf-2.65/html_node/Cache-Files.html):

> It is wrong to try to distribute cache files for particular system types. There is too much room for error in doing that, and too much administrative overhead in maintaining them. For any features that can't be guessed automatically, use the standard method of the canonical system type and linking files (see Manual Configuration).

> The site initialization script can specify a site-wide cache file to use, instead of the usual per-program cache. In this case, the cache file gradually accumulates information whenever someone runs a new configure script. (Running configure merges the new cache results with the existing cache file.) This may cause problems, however, if the system configuration (e.g., the installed libraries or compilers) changes and the stale cache file is not deleted.

So they say it's wrong to distribute these files, but I think Nix might be in a unique place where there is actually very little room for error in distributing them. Right?

If so, we could run a "dummy" `configure` once during the `stdenv` build (to build knowledge, with a ton of the standard checks turned on), and then automatically add `--config-cache ${config-cache}` to all our autoconf flags.

**Comment by vcunat** [comment]
Yes, but we would have to be extra careful for checks that might be invalidated by adding extra build inputs, setting env vars, etc.

**Comment by copumpkin** [comment]
@vcunat I was thinking of just having a `disableConfigCache` flag in `mkDerivation` that could turn it off if it causes trouble. There's a slight chance for things implicitly and quietly doing the wrong thing, but I don't think there are that many examples of flags we want to override in individual builds.

At the very least, the config.cache generator could have a "whitelist" of checks that we've decided are unconditionally (or extremely likely to be) safe.

**Comment by joachifm** [comment]
Looking at the sabotage config.cache, it's mostly type sizes and libc features. At least the type sizes ought to be safe, right?

**Comment by vcunat** [comment]
Yes, even all of what you list should be constant for a particular stdenv. We could

- find out what tests are often used,
- look them over manually and scratch any that might cause problems (when adding build inputs or flags or env vars),
- create our configure.ac with that list and use it to generate our cache for a stdenv,
- ensure this cache is used by default during configurePhase.

**Comment by peti** [comment]
The problem is that there are no rules whatsoever how cache entries are constructed, i.e. everyone is free to choose pretty much any key and assign it pretty much any value as he or she pleases. This makes cache files inherently local to the project, i.e. you're likely to run into conflicts when copying a cache file from project `foo` into project `bar` and assume that both use the same cache values meaning the same thing. That is *probably* true for cache entries created by standard Autoconf macros like `AC_PROG_CC` etc., but anything beyond that I would rather not rely on. So this raises the question of how effective can a shared `config.cache` be if all it contains are results from super elementary tests?

**Comment by joachifm** [comment]
The sabotage cache file is said to cut down 50% of tests, all of which look elementary.
I guess someone needs to implement a prototype for this feature and provide some numbers, or find some way to estimate how much time is being spent re-calculating the values we might think to cache.

I'd think, that with the number of builds, even minor savings would add up rather quickly.

**Comment by vcunat** [comment]
@peti: I certainly meant to include only the standard macros.

**Comment by spacekitteh** [comment]
Was anything ever done about this?

**Comment by vcunat** [comment]
I don't think so.  In general, I feel we have very many enhancement ideas relative to the amount we can realistically manage to actually implement.  It's one of factors contributing to our ever-growing "issue" list.

**Comment by Profpatsch** [comment]
Is it okay to close this? I don’t think this will go anywhere in the future.

**Comment by catern** [comment]
I know this is closed, but I think this would be a good first issue for someone contributing to Nixpkgs beyond packaging. Extensive benchmarking or great effort shouldn't be necessary - even a single entry in config.cache will speed up many builds, and it's just a matter of creating that file in the appropriate location.

**Comment by spacekitteh** [comment]
I'm kinda tempted to do it, all these years later, lol

**Comment by musteresel** [comment]
I already had a working - though probably not up to current Nixpkgs code quality standards - implementation in mid 2017; I try to find it.

**Comment by Profpatsch** [comment]
Well then, let’s get it rolling again!

**Comment by elikoga** [comment]
This looks interesting and I might want to try at implementing it sometime
Just commenting so I can find the issue later
